### PR TITLE
Add SwiftKotlin & improve Parse

### PR DESCRIPTION
**Pull request by inamiy**
Hi 😄
I added **Kotlin support** for this project which is inspired from https://github.com/angelolloqui/SwiftKotlin and obviously, [JavaプログラマのためのKotlin入門 - Qiita](http://qiita.com/koher/items/bcc58c01c6ff2ece658f) .
Sorry for pushing a large commit, but hopefully this will be interesting to you.

I also attached a sample playground file.

- Added Kotlin support (SwiftKotlin)
- Added playground file
- [Parse] Added `Struct`, `Protocol`, ~~`StringInterpolationLiteral`~~ `InterpolatedStringLiteral`

---

### 2017/05/24 21:43 Update

Here's my LT slide: [忙しいSwiftエンジニアのためのKotlin入門 // Speaker Deck](https://speakerdeck.com/inamiy/mang-siiswiftenziniafalsetamefalsekotlinru-men)

### CODE REVIEW

File: Sources/SwiftScript/Utilities.swift [hunk 1]
```diff
@@ -1,6 +1,16 @@
 import SwiftAST
 
-internal struct UnimplementedError: Error {}
+internal struct UnimplementedError: Error {
+    let functionName: StaticString
+    let fileName: StaticString
+    let lineNumber: Int
+
+    init(func functionName: StaticString, file fileName: StaticString, line lineNumber: Int) {
```
Comment (rintaro) [inline Sources/SwiftScript/Utilities.swift]: Why not
```swift
init(func functionName: StaticString = #function, file fileName: StaticString = #file, line lineNumber: Int = #line) {
```
then, on caller site, just:
```
throw UnimplementedError()
```
Doesn't it work?

**Review by rintaro — COMMENTED** [review]
Parser changes LGTM 👍

File: Sources/SwiftScript/Utilities.swift (hunk 1 above)
Comment (inamiy) [reply Sources/SwiftScript/Utilities.swift]: Thanks for check! Improved in 4a55650 ✨

File: Sources/SwiftAST/Literals.swift [hunk 2]
```diff
@@ -42,6 +42,14 @@ public struct StringLiteral: Literal {
     }
 }
 
+public struct StringInterpolationLiteral: Literal {
```
Comment (koher) [inline Sources/SwiftAST/Literals.swift]: Basically we have decided the type names of AST nodes following [the language reference](https://developer.apple.com/library/content/documentation/Swift/Conceptual/Swift_Programming_Language/LexicalStructure.html#//apple_ref/swift/grammar/string-literal) in "The Swift Programming Language".  So `InterpolatedStringLiteral` should be better.

Also I think we should introduce a new "StringLiteral" protocol with the "StaticStringLiteral" and the "InterpolatedStringLiteral" as concrete subtypes of it.

File: Sources/SwiftAST/Declarations.swift [hunk 3]
```diff
@@ -21,23 +21,29 @@ public struct ConstantDeclaration: Declaration {
...
     public var name: String
     public var type: Type_?
     public var expression: Expression?
-    // TODO: Computed Property
+    public var getSet: GetSet?  // TODO: Split to ProtocolPropertyDeclaration
+
+    public enum GetSet: Declaration {
+        case get
+        case getSet
+    }
```
Comment (koher) [inline Sources/SwiftAST/Declarations.swift:36]: I think it's OK so far. I think it should follow [this](https://developer.apple.com/library/content/documentation/Swift/Conceptual/Swift_Programming_Language/Declarations.html#//apple_ref/swift/grammar/getter-setter-keyword-block) later when we complete _properties_.

File: Sources/SwiftAST/Declarations.swift [hunk 4]
```diff
@@ -88,8 +94,15 @@ public struct EnumDeclaration: Declaration {
...
+    public var name: String
+    public var superTypes: [Type_]
+    public var members: [Declaration]
+
+    public init(name: String, superTypes: [Type_], members: [Declaration]) {
+        self.name = name
+        self.superTypes = superTypes
+        self.members = members
+    }
 }
```
Comment (koher) [inline Sources/SwiftAST/Declarations.swift:106]: Same as `GetSet`. ([reference](https://developer.apple.com/library/content/documentation/Swift/Conceptual/Swift_Programming_Language/Declarations.html#//apple_ref/swift/grammar/struct-declaration))

File: Sources/SwiftAST/Declarations.swift [hunk 5]
```diff
@@ -108,10 +121,13 @@ public struct ClassDeclaration: Declaration {
...
+    public var superTypes: [Type_]
+    public var members: [Declaration]
     
-    public init(name: String) {
+    public init(name: String, superTypes: [Type_], members: [Declaration]) {
         self.name = name
+        self.superTypes = superTypes
+        self.members = members
     }
 }
```
Comment (koher) [inline Sources/SwiftAST/Declarations.swift:132]: Same as `GetSet`. ([reference](https://developer.apple.com/library/content/documentation/Swift/Conceptual/Swift_Programming_Language/Declarations.html#//apple_ref/swift/grammar/protocol-declaration))

File: Sources/SwiftAST/Expressions.swift [hunk 6]
```diff
@@ -70,8 +70,12 @@ public struct FunctionCallExpression: Expression {
...
 public struct InitializerExpression: Expression {
-    // Unsupported
-    public init() {}
+    /// e.g. `SelfExpression`, `SuperclassExpression`.
+    public var receiverExpression: Expression
+
+    public init(receiverExpression: Expression) {
+        self.receiverExpression = receiverExpression
+    }
 }
```
Comment (koher) [inline Sources/SwiftAST/Expressions.swift]: Same as `GetSet`. (reference [1](https://developer.apple.com/library/content/documentation/Swift/Conceptual/Swift_Programming_Language/Expressions.html#//apple_ref/swift/grammar/initializer-expression), [2](https://developer.apple.com/library/content/documentation/Swift/Conceptual/Swift_Programming_Language/Expressions.html#//apple_ref/swift/grammar/self-initializer-expression))

**Review by koher — COMMENTED** [review]
Thank you for your great PR especially to support Kotlin and improve `JavaScriptTranslator` 😃

Although it is better that AST nodes follow [the language reference](https://developer.apple.com/library/content/documentation/Swift/Conceptual/Swift_Programming_Language/AboutTheLanguageReference.html#//apple_ref/doc/uid/TP40014097-CH29-ID345), I think it is so far OK. We have a lot to modify about AST independently on this PR.

File: Sources/SwiftAST/Literals.swift (hunk 2 above)
Comment (inamiy) [reply Sources/SwiftAST/Literals.swift]: Agree! I renamed the type in 49bd25e.

File: Sources/SwiftAST/Expressions.swift (hunk 6 above)
Comment (inamiy) [reply Sources/SwiftAST/Expressions.swift]: Renamed in de95f38.

**Comment by inamiy** [comment]
@koher @rintaro 
Thanks for review! I will remain others for next PR. Is that OK?

**Comment by koher** [comment]
It's OK for me👍

**Comment by rintaro** [comment]
LGTM, merging! 👍

**Pull request merged**